pkg/scheduler/plugins/capacity/capacity.go

```diff
@@ -220,133 +436,363 @@ func (cp *capacityPlugin) OnSessionOpen(ssn *framework.Session) {
+		childAttr.deserved.MinDimensionResource(childAttr.request, api.Zero)
 
-			klog.V(4).Infof("Capacity AllocateFunc: task <%v/%v>, resreq <%v>,  share <%v>",
-				event.Task.Namespace, event.Task.Name, event.Task.Resreq, attr.share)
-		},
-		DeallocateFunc: func(event *framework.Event) {
-			job := ssn.Jobs[event.Task.Job]
-			attr := cp.queueOpts[job.Queue]
-			attr.allocated.Sub(event.Task.Resreq)
-			metrics.UpdateQueueAllocated(attr.name, attr.allocated.MilliCPU, attr.allocated.Memory)
+		childAttr.deserved = helpers.Max(childAttr.deserved, childAttr.guarantee)
+		totalDeserved.Sub(oldDeserved).Add(childAttr.deserved)
+	}
 
-			cp.updateShare(attr)
+	// Check if the parent queue's deserved resources are less than the total deserved resources of child queues
+	if attr.deserved.LessPartly(totalDeserved, api.Zero) {
+		return fmt.Errorf("deserved resources of queue <%s> are less than the sum of its child queues' deserved resources", attr.name)
+	}
 
-			klog.V(4).Infof("Capacity EvictFunc: task <%v/%v>, resreq <%v>,  share <%v>",
-				event.Task.Namespace, event.Task.Name, event.Task.Resreq, attr.share)
-		},
+	// Check if the parent queue's guarantee resources are less than the total guarantee resources of child queues
+	if attr.guarantee.LessPartly(totalGuarantee, api.Zero) {
+		return fmt.Errorf("guarantee resources of queue <%s> are less than the sum of its child queues' guarantee resources", attr.name)
+	}
+
+	for _, childAttr := range attr.children {
+		err := cp.checkHierarchicalQueue(childAttr)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (cp *capacityPlugin) createOrUpdateRootQueue(vcClient vcclient.Interface) error {
+	queue, err := vcClient.SchedulingV1beta1().Queues().Get(context.TODO(), cp.rootQueue, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			queue = &vcv1beta1.Queue{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: string(cp.rootQueue),
+				},
+				Spec: vcv1beta1.QueueSpec{},
+			}
+			cp.makeRootQueue(queue)
+
+			return retry.OnError(wait.Backoff{
+				Steps:    5,
+				Duration: time.Second,
+				Factor:   1,
+				Jitter:   0.1,
+			}, func(err error) bool {
+				return err != nil && !apierrors.IsAlreadyExists(err)
+			}, func() error {
+				_, err := vcClient.SchedulingV1beta1().Queues().Create(context.TODO(), queue, metav1.CreateOptions{})
+				return err
+			})
+		}
+
+		return err
+	}
+
+	cp.makeRootQueue(queue)
+	return retry.OnError(wait.Backoff{
+		Steps:    5,
+		Duration: time.Second,
+		Factor:   1,
+		Jitter:   0.1,
+	}, func(err error) bool {
+		return err != nil
+	}, func() error {
+		_, err := vcClient.SchedulingV1beta1().Queues().Update(context.TODO(), queue, metav1.UpdateOptions{})
+		return err
 	})
 }
 
-func (cp *capacityPlugin) OnSessionClose(ssn *framework.Session) {
-	cp.totalResource = nil
-	cp.totalGuarantee = nil
-	cp.queueOpts = nil
+func (cp *capacityPlugin) makeRootQueue(queue *vcv1beta1.Queue) {
```

Comment:
Why do we need to update `rootQueue.Deserved` and `rootQueue.Guarantee` here?
It seems that we will do
```go
	if attr.name == cp.rootQueue {
		attr.guarantee = totalGuarantee
		cp.totalGuarantee = totalGuarantee
		attr.realCapability = cp.totalResource
		attr.deserved = cp.totalResource
	}
```
for root queue in each OpenSession -> `checkHierarchicalQueue()`

Reply:
The logic in `checkHierarchicalQueue()` is to update the `queueAttr` of the root queue, which is mainly to preserve the attributes of the root queue during the calculation process of the capacity plugin. The `makeRootQueue` function is to supplement the attributes in the root queue CR, thereby updating the queue CR.

Reply:
Ok, it's good for other plugins to use queue CR in the future :laughing: